### ISSUE TITLE
Android APK fails to run due to NullPointerException ConstructPlayGames.java

### ISSUE THREAD

**Issue by sotano42** [Scirra/Construct-3-bugs#1848]
## Problem description

I have exported my game using the APK build service of construct and the game and published application is causing NullPointerException terminations on several devices, I published games before with older versions of C3 without problems. Devices found to crash are the following:

- Aiwa AW790 (AW790), Android 7.0
- Motorola Moto G (5th Gen) (cedric), Android 7.0
- ZTE Z971 (peony), Android 7.1
- Motorola Moto G (1st Gen) (falcon_umts), Android 5.1
- HTC One (M8) (htc_m8), Android 6.0
- Huawei Y5 2017 (HWMYA-L6737), Android 6.0
- Huawei Y5II (HWCUN-L6735), Android 5.1
- Samsung Galaxy J7 (j7elte), Android 6.0

All of those devices start the game and end up with a blank screen sometimes just playing the sound and receiving input from the touch plugin.

## Attach a .c3p

The project attached is not a full project, is a version with one level and 100 events just for you to test. I tried removing plugins and functionality but couldn't find a clue of what's happening. I couldn't isolate the cause.

[Bronco Wave (8).zip](https://github.com/Scirra/Construct-3-bugs/files/2228454/Bronco.Wave.8.zip)

## Steps to reproduce

Compile, sign and install on any of the listed devices, application won't run.

## Observed result

Application loads with a blank screen sometimes unresponsive, sometimes with touch plugin working but always with a blank screen.
![image](https://user-images.githubusercontent.com/10860045/43210644-42eb531e-8ff5-11e8-9178-bf64f3a739a4.png)

## Expected result

Application must run.

## Affected browsers

- Aiwa AW790 (AW790), Android 7.0
- Motorola Moto G (5th Gen) (cedric), Android 7.0
- ZTE Z971 (peony), Android 7.1
- Motorola Moto G (1st Gen) (falcon_umts), Android 5.1
- HTC One (M8) (htc_m8), Android 6.0
- Huawei Y5 2017 (HWMYA-L6737), Android 6.0
- Huawei Y5II (HWCUN-L6735), Android 5.1
- Samsung Galaxy J7 (j7elte), Android 6.0

## System details

<details><summary>The following dump refers to ConstructPlayGame.java component, info was fetched from Android vitals</summary>

java.lang.RuntimeException:
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:3392)
  at android.app.ActivityThread.handleResumeActivity (ActivityThread.java:3423)
  at android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:2725)
  at android.app.ActivityThread.access$900 (ActivityThread.java:196)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1538)
  at android.os.Handler.dispatchMessage (Handler.java:111)
  at android.os.Looper.loop (Looper.java:210)
  at android.app.ActivityThread.main (ActivityThread.java:5988)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:852)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:742)

Caused by: java.lang.RuntimeException: 
  at android.app.ActivityThread.deliverResults (ActivityThread.java:4013)
  at android.app.ActivityThread.performResumeActivity (ActivityThread.java:3373)

Caused by: java.lang.NullPointerException: 
  at com.scirra.ConstructPlayGames.updateSignInState (ConstructPlayGames.java:191)
  at com.scirra.ConstructPlayGames.onDisconnected (ConstructPlayGames.java:236)
  at com.scirra.ConstructPlayGames.onActivityResult (ConstructPlayGames.java:90)
  at org.apache.cordova.CordovaInterfaceImpl.onActivityResult (CordovaInterfaceImpl.java:159)
  at org.apache.cordova.CordovaActivity.onActivityResult (CordovaActivity.java:359)
  at android.app.Activity.dispatchActivityResult (Activity.java:6674)
  at android.app.ActivityThread.deliverResults (ActivityThread.java:4009)

</details>

**Comment by shortercode** [comment]
I think this is caused by the activity swap under low memory conditions, it's going to require a little tweaking to get the plugin to work correctly under this situation.

Thanks for the report.

**Comment by shortercode** [comment]
Looking closer at your report I inclined to think you have 2 separate issues here. I would expect the app to exit with the stack trace you have sent me, not produce a black screen.

There's a known graphics bug doing the rounds on Android at the moment which has been causing problems for us, we published a workaround in r110 which should prevent the "vanishing game" bug. If you export with r110+ do you see that issue?

What percentage of sessions are you seeing the null pointer exception crash on?

**Comment by sotano42** [comment]
I'm not completely sure the problem is related to low memory conditions since some of the devices that failed hare quite decent and the failure is constant, not sporadic. Considering that my app has just a few installs from 470 installs I have 15 dumps (3% failure) reported on android vitals plus some fails I saw on the devices of some friends that are not showing up in android vitals, so I'm unsure on how many devices really couldn't run the game.

For what I saw it looks like I was on r104, I'm compiling a new APK and let me contact some of my friends since I don't have a device that fails. Will come back to you as soon as I have feedback.

**Comment by sotano42** [comment]
Ok, so compiling on version r111 worked and the app ran again correctly but in device **Huawei A7 2017** it can't play any sound. I'm not getting any dump yet, will try to investigate further but at least two of the failed devices now were able to run the game.

**Comment by shortercode** [comment]
I'm glad to hear r111 solved your major problem, that graphics issue has been affecting a lot of our users so it's good that a workaround was found. We're not sure when the underlying bug will be fixed, as we haven't found a reliable simple replication for the chrome team yet.

I reviewed the code around the null pointer exception and the only way I could see the value being null is if the main activity had been released while a sign in dialog was being shown. This is something Android *can* do, but should only do it in low memory situations.

I've rewritten the code around the sign in logic to avoid the crash, although the exact behaviour of the sign in flow may be a little odd. Not a huge amount that can be done about that though, it's a quirk of Android development...

Not really sure what the Huawei audio issue is but I expect it's lacking opus support for some reason. I would check to see that it's chrome version is up to date.
